system/Image/Image.php

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * CodeIgniter
+ *
+ * An open source application development framework for PHP
+ *
+ * This content is released under the MIT License (MIT)
+ *
+ * Copyright (c) 2014 - 2016, British Columbia Institute of Technology
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @package	CodeIgniter
+ * @author	CodeIgniter Dev Team
+ * @copyright	Copyright (c) 2014 - 2016, British Columbia Institute of Technology (http://bcit.ca/)
+ * @license	http://opensource.org/licenses/MIT	MIT License
+ * @link	http://codeigniter.com
+ * @since	Version 3.0.0
+ * @filesource
+ */
+
+/**
+ * Image manipulation factory.
+ *
+ * Creates and returns an instance of the appropriate image manipulation processor.
+ *
+ * @package CodeIgniter\Image
+ */
+class Image
+{
+    protected $driver;
+    protected $resource;
+    protected $imageFullPath;
+    protected $imageFolder;
+    protected $imageFilename;
+    protected $imageExtension;  
+    protected $mime;
+    
+    protected $imageTypes = [
+            'jpg',
+            'jpeg',
+            'png',
+            'gif',
+            'tiff'
+    ];
+    
+    // ------------------------------------------------------------------
+    
+    /**
+     * @param string $driver
+     * @return resource
+     */
+    public function load(string $driver): resource
+    {
+        $classname = "Codeigniter\\Image\\$driver";
+        return new $classname;
+    }
+    
+    // ------------------------------------------------------------------
+    
+    /**
+     * @param string $folder
+     * @return resource
+     */
+    public function setFolder(string $folder): resource
+    {
+        $this->imageFolder = $folder;
+        return $this;
+    }
+    
+    // ------------------------------------------------------------------
+    
+    /**
+     * @param string $filename
+     * @return resource
+     */
+    public function setFilename(string $filename): resource
```

Comment:
The `setFolder()`, `setFilename()`, and similar methods seem like they would be better attached to the different Handlers as ways to set new attributes on the files. Actually, that doesn't make sense, either, because you'd pass that information directly to the `copy()` command. Since the `image()` method below already takes a full path, I'd say get rid of these functions. This also frees this class up to be a little more useful than just a factory. I'll make a note of that elsewhere.